compiler/src/iree/compiler/Dialect/Flow/Transforms/Canonicalizer.cpp

```diff
@@ -18,80 +20,63 @@ namespace mlir::iree_compiler::IREE::Flow {
-  using OpRewritePattern<tensor::PadOp>::OpRewritePattern;
-
-  LogicalResult matchAndRewrite(tensor::PadOp padOp,
+static std::optional<SmallVector<OpFoldResult>> getDefiningMixedSizes(Value v) {
+  if (auto empty = v.getDefiningOp<tensor::EmptyOp>()) {
+    return empty.getMixedSizes();
+  } else if (auto extract = v.getDefiningOp<tensor::ExtractSliceOp>()) {
+    // TODO: Support rank reducing cases.
+    if (extract.getSourceType().getRank() !=
+        extract.getResultType().getRank()) {
+      return {};
+    }
+    return extract.getMixedSizes();
+  }
+  return {};
+}
+
+struct FoldFullInsertSlice : public OpRewritePattern<tensor::InsertSliceOp> {
+  using OpRewritePattern<tensor::InsertSliceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::InsertSliceOp insertSliceOp,
                                 PatternRewriter &rewriter) const override {
-    if (padOp.getNofold()) {
-      return failure();
+    if (!insertSliceOp.hasUnitStride() || !insertSliceOp.hasZeroOffset()) {
+      return rewriter.notifyMatchFailure(insertSliceOp,
+                                         "non-unit stride or non-zero offset.");
     }
-    auto producerPad = padOp.getSource().getDefiningOp<tensor::PadOp>();
-    if (!producerPad || producerPad.getNofold()) {
+
+    RankedTensorType sourceType = insertSliceOp.getSourceType();
+    RankedTensorType resultType = insertSliceOp.getResultType();
+    if (sourceType != resultType) {
       return rewriter.notifyMatchFailure(
-          padOp, "producer is not a foldable tensor.pad op");
+          insertSliceOp,
+          "unimplemented: Cast-like or reshape-like insert ops.");
     }
 
-    // Fail if the tensor::PadOps padding values do not match.
-    Value consumerPadValue = padOp.getConstantPaddingValue();
-    Value producerPadValue = producerPad.getConstantPaddingValue();
-    if (!consumerPadValue || !producerPadValue ||
-        consumerPadValue != producerPadValue) {
+    std::optional<SmallVector<OpFoldResult>> mixedSizes =
+        getDefiningMixedSizes(insertSliceOp.getDest());
+    if (!mixedSizes) {
       return rewriter.notifyMatchFailure(
-          padOp, "cannot fold PadOps with different padding values");
+          insertSliceOp, "Could not find producer with list of tensor sizes.");
     }
 
-    Location loc = padOp.getLoc();
-    AffineExpr d0, d1;
-    bindDims(rewriter.getContext(), d0, d1);
-
-    // Combine the low/high paddings of the two tensor::PadOps.
-    auto addPaddings = [&](ArrayRef<OpFoldResult> consumerPaddings,
-                           ArrayRef<OpFoldResult> producerPaddings) {
-      SmallVector<OpFoldResult> sumPaddings;
-      for (auto [consumerIndex, producerIndex] :
-           llvm::zip_equal(consumerPaddings, producerPaddings)) {
-        sumPaddings.push_back(affine::makeComposedFoldedAffineApply(
-            rewriter, loc, d0 + d1, {consumerIndex, producerIndex}));
+    for (auto [insertSize, destSize] :
```

Comment:
I think `OpFoldResult` has equality function that does this for you.

Reply:
Unfortunately different `getMixedSizes` implementations returns different attribute types `(i64 vs index)`: https://discord.com/channels/689900678990135345/689906000043573354/1301292218312822876

I'll add a comment explaining why here.